src/storage/azure/storage.c

```diff
@@ -363,35 +361,61 @@ storageAzureListInternal(
                 queryPrefix = strNewFmt("%s%s", strZ(basePrefix), strZ(expressionPrefix));
         }
 
-        // Loop as long as a continuation token returned
+        // Create query
+        HttpQuery *query = httpQueryNewP();
+
+        // Add the delimiter to not recurse
+        if (!recurse)
+            httpQueryAdd(query, AZURE_QUERY_DELIMITER_STR, FSLASH_STR);
+
+        // Add resource type
+        httpQueryAdd(query, AZURE_QUERY_RESTYPE_STR, AZURE_QUERY_VALUE_CONTAINER_STR);
+
+        // Add list comp
+        httpQueryAdd(query, AZURE_QUERY_COMP_STR, AZURE_QUERY_VALUE_LIST_STR);
+
+        // Don't specify empty prefix because it is the default
+        if (!strEmpty(queryPrefix))
+            httpQueryAdd(query, AZURE_QUERY_PREFIX_STR, queryPrefix);
+
+        // Loop as long as a continuation marker returned
+        HttpRequest *request = NULL;
+
         do
         {
             // Use an inner mem context here because we could potentially be retrieving millions of files so it is a good idea to
             // free memory at regular intervals
             MEM_CONTEXT_TEMP_BEGIN()
             {
-                HttpQuery *query = httpQueryNewP();
-
-                // Add continuation token from the prior loop if any
-                if (marker != NULL)
-                    httpQueryAdd(query, AZURE_QUERY_MARKER_STR, marker);
+                // Get sync response on first request or async response on continuation
+                HttpResponse *response = NULL;
 
-                // Add the delimiter to not recurse
-                if (!recurse)
-                    httpQueryAdd(query, AZURE_QUERY_DELIMITER_STR, FSLASH_STR);
+                if (request == NULL)
+                    response = storageAzureRequestP(this, HTTP_VERB_GET_STR, .query = query);
```

Comment:
I'm a little confused why line 394 would not be wrapped in storageAzureResponseP. If line 397 is 
`response = storageAzureResponseP(request); `
then response should be coming from a call to a response function. By line 397, `request` was set by line 415:
`request = storageAzureRequestAsyncP(this, HTTP_VERB_GET_STR, .query = query);`

then why wouldn't line 394 be:
`response = storageAzureResponseP(storageAzureRequestP(this, HTTP_VERB_GET_STR, .query = query));`

I mean is `response` a response or is it a request?  That's the part that is a bit confusing - what really is the variable `response`?

Reply:
`storageAzureResponseP()` is used to get a response from a request started asynchronously by `storageAzureRequestAsyncP()`. storageAzureRequestP() does an sync request and returns the response immediately.

If the request is NULL it means we don't have any date yet, so we need to do a sync request to get some. Doing an async request here does us not good since we don't have anything else we can do while waiting for the data.

> what really is the variable response

It's defined as HttpResponse on line 391.

Is the confusion here over the naming of the functions, perhaps? We could work on that, but as part of another commit.

Reply:
Fixed some typos in the comment above, perhaps you should re-read.

Reply:
It's definitely a naming thing. I think at a minimum, could you add a couple comments: 
at line 394: 
`get an immediate response from a synchronous request`
and at line 397:
`get a response from a request started asynchronously`

Reply:
I kind of thought that's what this comment said:
```
// Get sync response on first request or async response on continuation
```
Do you want to expand on that comment, move it, or just say it again in the if-else?

Reply:
How about:
```
                // If there is no outstanding async request then get the response immediately from a sync request
                HttpResponse *response = NULL;

                if (request == NULL)
                {
                    response = storageS3RequestP(this, HTTP_VERB_GET_STR, FSLASH_STR, query);
                }
                // Else get the response from the outstanding sync request
                else
                {
                    response = storageS3ResponseP(request);

                    httpRequestFree(request);
                    request = NULL;
                }
```

Reply:
Or maybe it is clearer to invert it:
```
                // If there is an outstanding async request then wait for the response
                HttpResponse *response = NULL;

                if (request != NULL)
                {
                    response = storageS3ResponseP(request);

                    httpRequestFree(request);
                    request = NULL;
                }
                // Else get the response immediately from a sync request
                else
                {
                    response = storageS3RequestP(this, HTTP_VERB_GET_STR, FSLASH_STR, query);
                }
```

Reply:
If you locate 
`// If there is no outstanding async request then get the response immediately from a sync request`
right above 
`if (request == NULL)`
then that would help. Proximity of the comment is important here. The naming of the functions is the issue and I know I should have caught it before but that was the PR I had a bit of trouble following. As you said, we can make adjustments to the function names in another commit. And I KNOW naming things is HARD! :-)

Reply:
OK, done that way.